opal/mca/base/mca_base_source.c

```diff
@@ -0,0 +1,238 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil -*- */
+/*
+ * Copyright (c) 2013-2018 Los Alamos National Security, LLC. All rights
+ *                         reserved.
+ * Copyright (c) 2015      Cisco Systems, Inc.  All rights reserved.
+ * Copyright (c) 2015      Bull SAS.  All rights reserved.
+ * Copyright (c) 2015      The University of Tennessee and The University
+ *                         of Tennessee Research Foundation.  All rights
+ *                         reserved.
+ * Copyright (c) 2017      IBM Corporation. All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include "mca_base_source.h"
+#include "mca_base_vari.h"
+
+#include <stddef.h>
+#include <sys/time.h>
+#include <sys/resource.h>
+
+#include "opal/class/opal_pointer_array.h"
+#include "opal/class/opal_hash_table.h"
+
+static opal_pointer_array_t registered_sources;
+static int mca_base_source_initialized;
+static int source_count;
+
+int mca_base_source_default_source = -1;
+
+static uint64_t mca_base_source_default_time_source (void)
+{
+    uint64_t time_value;
+
+#if OPAL_HAVE_CLOCK_GETTIME
+    struct timespec current;
+
+    clock_gettime (CLOCK_MONOTONIC, &current);
```

Comment:
A larger question: given that MPI_WTICK() and MPI_TIME() exist, should MPI_T events use the same back-end time handling?  And if so, should that be based on the OPAL timer framework?

For the second question, it looks like Open MPI has a complicated history regarding time/Wtime/Wtick/the OPAL timer framework.  See, for example:

* https://github.com/open-mpi/ompi/issues/3003
* https://github.com/open-mpi/ompi/wiki/Timers

There was a lot of discussion about RTDSC timers, frequency scaling, whether RTDSC was monotonic, etc.  Currently (Dec 2021), Wtick/Wtime:

1. On Linux, use `clock_gettime()` (not the OPAL timer framework)
2. Use `gettimeofday()` everywhere else.

Is there a reason we're not using `clock_gettime()` everywhere it's available?  (and potentially `CLOCK_MONOTONIC_RAW` if it's available)?  This may be a naive question -- I haven't done any perf testing of `CLOCK_MONOTONIC_RAW` on Linux, for example.

FWIW, the RHEL 6 `clock_gettime(2)` says this:

```
       CLOCK_MONOTONIC
              Clock  that  cannot be set and represents monotonic time since
              some unspecified starting point.  This clock is  not  affected
              by discontinuous jumps in the system time (e.g., if the system
              administrator manually changes the clock), but is affected  by
              the incremental adjustments performed by adjtime(3) and NTP.

       CLOCK_MONOTONIC_RAW (since Linux 2.6.28; Linux-specific)
              Similar to CLOCK_MONOTONIC, but provides access to a raw hard-
              ware-based  time that is not subject to NTP adjustments or the
              incremental adjustments performed by adjtime(3).
```

which directly states that `CLOCK_MONOTONIC` is affected by `adjtime(3)` and NTP, which means we probably shouldn't be using it anywhere we expect actually-monotonic time.  Instead, it seems we should be using `CLOCK_MONOTONIC_RAW`...?  That being said, I don't know if `CLOCK_MONOTONIC_RAW` is subject to a per-hardware-core value, and may change if your process' locality changes to a different core.  This would need to be verified.

Reply:
switched to using the opal_clock facility